tools/hatari-prg-profile.1
.TH "hatari-prg-profile" "1" "2026-10-18" "Hatari" "Hatari utilities"
.SH NAME
hatari\-prg\-profile \- profile performance of given Atari program
.SH SYNOPSIS
.B hatari\-prg\-profile
<\fIpath\fP> <\fIopcode\fP> [\fIoptions\fP]
.PP
.B hatari\-prg\-profile
<\fIreport\fP>
.SH DESCRIPTION
Runs given Atari program from GEMDOS HD with Hatari and profiles its
execution until program calls the specified GEMDOS \fIopcode\fP. Saves
the profile data, post-processes it, and shows overview of the
generated profile report.
.PP
If given existing profiling \fIreport\fP (not necessarily produced by
this utility), shows a summary of it.
.PP
Profiling requires program to have debug symbols; either included to
the binary itself, or in a separate "<program name>.sym" file.
.PP
Profile data is saved to same directory where the Atari program is,
and the report generated from it, is saved to the current directory.
.SH ARGUMENTS
.TP
.B path
Path to Atari program to profile.
.TP
.B opcode
2-digit GEMDOS opcode in hex.
.TP
.B options
Extra options for provide for Hatari.
.SH EXAMPLES
.TP
Profile program until it terminates with Pterm() (0x4C):
.B hatari\-prg\-profile ./hello.tos 4C
.TP
Profile OPL2 DSP midiplayer until Fcreate() (0x3C), under Falcon emulation:
.B hatari\-prg\-profile ./f030mid.tos 3C \-\-machine falcon \-s 4
.SH FILES
After succesful profiling run, current directory includes following
files:
.TP
.I profile-data.txt
Profile disassembly and symbol callgraph information, saved at the end
of the profiling.  Disassembly is annotated with per-address counts of
how many times given instruction was run and how many cycles that
used.
.IP
For machines with emulated cache, address annotations include also
number of instruction cache misses and data cache hits.
.TP
.I profile-report.txt
Profiling report generated by post-processing the
\fIprofile-data.txt\fP file. Shows how long the profile run took
(which can be used for timing program performance), and sorted lists
of which program symbols were called most often, how many cycles
(time) code under those took etc.
.SH NOTES
While whole program profiling can be useful by itself (depending on
program), this script can also be used e.g. as a starting point for
automating profiling of more specific parts of the program. That would
require (at least) setting up suitable (program-specific) breakpoints
(to start profiling and save the data) for relevant points in its
execution.
.PP
Script does not generate callgraphs because for more complex programs
finding suitable post-processing options for reducing callgraph to a
readable size, while still retaining relevant information, can take
quite a lot of iteration.
.PP
In simple programs though, only thing needed for reasonable
callgraphs, is re-running the post-processor command (shown by the
script) with an additional "\-g" option, and viewing the resulting
\fIGraphviz\fP ".dot" files with (e.g.) \fIxdot\fP utility.
.PP
For more info, see: https://www.hatari\-emu.org/doc/debugger.html#Profiling
.SH SEE ALSO
.IR hatari (1), xdot (1)
.SH "AUTHOR"
Written by Eero Tamminen <oak at helsinkinet fi>.
.SH "LICENSE"
This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or (at
your option) any later version.
.SH "NO WARRANTY"
This program is distributed in the hope that it will be useful, but
WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
General Public License for more details.


tools/hatari-prg-profile.sh
#!/bin/sh

# or override these with paths to non-installed Hatari version
hatari=$(which hatari)
profile=$(which hatari_profile)
gst2ascii=$(which gst2ascii)

# show profile report summary: $1=report
show_overview ()
{
	awk '
	/^CPU profile/ { out=1 }
	/^Time / { print $0 "\n"; out=0 }
	/^Used / { out=1 }
	/^Instruction/ { exit }
	/^ *0\./ { if(out) { print; exit } }
	{ if (out) print }
	' "$1"
	echo
}

# post_process profile: <symbols file> <profile file> <report file>
post_process ()
{
	symbols=$1
	input=$2
	output=$3

	echo "$profile -stp -r $symbols --limit 0.1 $input > $output"
	$profile -stp -r "$symbols" --limit 0.1 "$input" > "$output"

	echo
	ls -l "$input" "$output"
}

# show script help & exit: $1=errmsg
error_exit ()
{
	name=${0##*/}
	cat <<EOF 1>&2

Run given Atari program from GEMDOS HD with Hatari and profile its
execution until program calls specified GEMDOS opcode. Save the
profile data, post-processes it, and show overview of the generated
profile report.

Usage:
	$name <path> <opcode> [options]
	$name <report>

Arguments:
	path    - Path to Atari program to profile
	opcode  - 2-digit GEMDOS opcode in hex
	options - Extra options to provide for Hatari
	report  - Profiler report

Examples:
	# profile program until it terminates with Pterm() (0x4C):
	$name ./hello.tos 4C

	# profile OPL2 DSP midiplayer until Fcreate() (0x3C),
	# under Falcon emulation:
	$name ./f030mid.tos 3C --machine falcon -s 4

	# show overview of earlier produced profile report
	$name $data

ERROR: $1!

EOF
	exit 1
}

if [ $# -eq 1 ] && [ -f "$1" ]; then
	show_overview "$1"
	exit 0
fi

# process command line args
if [ $# -lt 2 ]; then
	error_exit "not enough arguments"
fi

prg=$1
if [ ! -e "$prg" ]; then
	error_exit "specified '$prg' Atari program to profile, is missing"
fi
shift

opcode=$1
if [ ${#opcode} -ne 2 ]; then
	error_exit "specified '$opcode' is not two-digit hex number"
fi
shift

hatari_args=$*

# verify required tools exist
if [ ! -x $hatari ]; then
	error_exit "'hatari' not found, is it in PATH?"
fi
if [ ! -x $profile ]; then
	error_exit "'hatari_profile' [.py] not found, is it in PATH?"
fi
if [ ! -x $gst2ascii ]; then
	error_exit "'gst2ascii' not found, is it in PATH?"
fi

# program dir
dir=${prg%/*}
if [ ! -d "$dir" ]; then
	error_exit "Atari program dir '$dir' missing"
fi

# program name without path
prg=${prg##*/}

# program name without extension
base=$(echo "$prg" | sed -E 's/[.](app|gtp|prg|tos|ttp|APP|GTP|PRG|TOS|TTP)//')

# check for / extract program symbols required by post-processing
syms="$dir/$base.sym"
if [ ! -f "$syms" ]; then
	echo "Warning: '$syms' symbols file missing, extracting them from the program for profile post-processing"
	# skip non-TEXT section symbols
	if ! $gst2ascii -a -b -d "$dir/$prg" > "$syms"; then
		error_exit "'$prg' does not include symbols, cannot profile"
	fi
	if grep -q " _ZN" "$syms"; then
		echo "Symbols seem to be C++ / mangled, demangling with host 'c++filt' (in binutils)..."
		c++filt < "$syms" > "$syms.tmp"
		mv "$syms.tmp" "$syms"
	fi
else
	echo "Using pre-existing symbols file '$syms'"
fi

# debugger scripts for profiling
start="profile-start.ini"
save="profile-save.ini"

# profile data save file
data="profile-data.txt"

# start program (CPU side) profiling, assumes symbols are already loaded)
# invoke second debugger script when program calls specified GEMDOS opcode
cat > "$start" << EOF
profile on
b GemdosOpcode = 0x$opcode :trace :file ${save##*/}
EOF

# save collected profile & set Hatari quit flag / exit value
cat > "$save" << EOF
profile save $data
quit 0
EOF

# show debugger scripts
head "$start" "$save"

# remove old profile data to later check whether it's creation succeeded
rm "$data"

# profile the program
# shellcheck disable=SC2086
$hatari --trace os_base,event --symload exec \
  --parse "prg:$start"  $hatari_args  "$dir/$prg"

# check that profiling succeeded to end
if [ ! -f "$data" ]; then
	error_exit "profiling failed, profile data file '$data' missing"
fi

# remove debugger scripts
rm "$start" "$save"

# generated report
report="profile-report.txt"
 
# post-process saved profile data
post_process "$syms" "$data" "$report"
echo

# show overview of the generated report
show_overview "$report"
